HeterogeneousCore/AlpakaCore/interface/alpaka/EDMetadata.h

```diff
@@ -0,0 +1,86 @@
+#ifndef HeterogeneousCore_AlpakaCore_interface_alpaka_EDMetadata_h
+#define HeterogeneousCore_AlpakaCore_interface_alpaka_EDMetadata_h
+
+#include <atomic>
+#include <memory>
+
+#include <alpaka/alpaka.hpp>
+
+#include "FWCore/Concurrency/interface/WaitingTaskWithArenaHolder.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/HostOnlyTask.h"
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+  /**
+   * The EDMetadata class provides the exact synchronization
+   * mechanisms for Event data products for backends with asynchronous
+   * Queue. These include
+   * - adding a notification for edm::WaitingTaskWithArenaHolder
+   * - recording an Event
+   * - synchronizing an Event data product and a consuming EDModule
+   *
+   * For synchronous backends the EDMetadata acts as an owner of the
+   * Queue object, as no further synchronization is needed.
+   *
+   * EDMetadata is used as the Metadata class for
+   * edm::DeviceProduct<T>, and is an implementation detail (not
+   * visible to user code).
+   */
+
+#ifdef ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED
+  // All backends with a synchronous queue
```

Comment:
@fwyzard I think all other uses of these `#ifdef`s to check if the backend's device is host, and/or if the queue is blocking, could be easily replaced with e.g. `if constexpr` and traits, except this one. I think it would be technically possible here as well, but I think it would be more cumbersome as the classes are quite different for the two case.